src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/AuditUtility.cs

```diff
@@ -78,10 +90,100 @@ public AuditUtility(
+                    List<PackageVulnerabilityInfo>? knownVulnerabilitiesForPackage = GetKnownVulnerabilities(downloadDependency.Name, downloadDependency.VersionRange.MaxVersion!, knownVulnerabilities);
+
+                    if (knownVulnerabilitiesForPackage?.Count > 0)
+                    {
+                        DownloadDependency downloadDependencyPackage = new(downloadDependency.Name, downloadDependency.VersionRange);
+
+                        foreach (PackageVulnerabilityInfo knownVulnerability in knownVulnerabilitiesForPackage)
+                        {
+                            if ((int)knownVulnerability.Severity < (int)MinSeverity && knownVulnerability.Severity != PackageVulnerabilitySeverity.Unknown)
+                            {
+                                continue;
+                            }
+
+                            if (SuppressedAdvisories?.TryGetValue(knownVulnerability.Url.OriginalString, out bool advisoryUsed) == true)
+                            {
+                                TotalPackageDownloadWarningsSuppressedCount++;
+
+                                if (!advisoryUsed)
+                                {
+                                    SuppressedAdvisories[knownVulnerability.Url.OriginalString] = true;
+                                    DistinctPackageDownloadAdvisoriesSuppressedCount++;
+                                }
+
+                                continue;
+                            }
+
+                            result ??= new();
+
+                            if (!result.TryGetValue(downloadDependency, out PackageDownloadAuditInfo? auditInfo))
+                            {
+                                auditInfo = new(downloadDependency);
+                                result.Add(downloadDependency, auditInfo);
+                            }
+
+                            if (!auditInfo.GraphsPerVulnerability.TryGetValue(knownVulnerability, out List<string>? affectedGraphs))
+                            {
+                                affectedGraphs = new();
+                                auditInfo.GraphsPerVulnerability.Add(knownVulnerability, affectedGraphs);
+                            }
+
+                            // Multiple package sources might list the same known vulnerability, so de-dupe those too.
+                            if (!affectedGraphs.Contains(downloadDependency.Name))
+                            {
+                                affectedGraphs.Add(downloadDependency.Name);
+                            }
+                        }
+                    }
+                }
+            }
+
+            return result;
+        }
+
+
         public async Task<bool> CheckPackageVulnerabilitiesAsync(CancellationToken cancellationToken)
         {
             // Performance: Early exit if restore graph does not contain any packages.
-            if (!HasPackages())
+            if (!HasPackages() && (_targetFrameworks.Count == 0))
```

Comment:
I don't think it's possible for there to ever be `_targetFrameworks.Count == 0`. Any project with a valid `<TargetFramework>` is going to have a non-empty collection, even if it doesn't have any PackageReference or PackageDownload items.

There is a test that attempts to check for this scenario: https://github.com/NuGet/NuGet.Client/blob/949031c79ea7f5c15b59f31e26d5355fb23795ca/test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests/Utility/AuditUtilityTests.cs#L109

However, due to how this class and the test helper methods were modified, it allows the test to use data that is impossible in production, and therefore fail to catch this regression.

There might be better ideas, but my first idea is to have code that validates that the list of `RestoreTargetGraph[]` TFMs match the `IList<TargetFrameworkInformation>` TFMs.  Both in count and actually validate that the two collections use identical TFMs. Since restore perf is critical, this test should be done in the test helper. While most tests in the class use the "context" class, Check_MultiTargetingProjectFile_WarningsHaveExpectedProperties does not use it, so it would be great if you could find a way to implement this validation that works for that test as well (or improve the "context" class so that it can be used by this test)

Reply:
I have updated the if statement to actually go through the TFMs and count `PackageDownload`s. This should not have perf regressions. Since we already have another loop that goes through all the `PackageDownload`s, adding one more loop for counting purposes should be okay as long as we don't do an allocation.

I have updated the test context to have at least one TFM.